SpatialGDK/Documentation/content/non-unreal-layers.md

```diff
@@ -0,0 +1,213 @@
+	* @param NextEntityId the next available entity ID in the snapshot, this reference should be incremented appropriately.
+	* @return bool the success of writing to the snapshot output stream, this is returned to the overall snapshot generation.
+	**/
+bool WriteToSnapshotOutput(Worker_SnapshotOutputStream* OutputStream, Worker_EntityId& NextEntityId);
+```
+
+## Example
+
+Below is a simple, example schema file which could be used by a separate worker layer to track player statistics:
+
+```
+package improbable.session;
+
+type MyRequest {
+  string player_name = 1;
+}
+type MyResponse {}
+
+component Session {
+    id = 1337;
+    uint32 player_count = 1;
+    command MyResponse some_command(My_Request);
+}
+```
+
+#### Sending data
+
+You could serialize and send a component update in user code in the following way:
+
+```
+void SendSomeUpdate(Worker_EntityId target_entity_id, Worker_ComponentId component_id) {
+    Worker_ComponentUpdate component_update;
+    component_update.id = component_id
+    component_update.schema_type = Schema_CreateComponentUpdate(component_id);
+
+    //  Component specific serialization
+    Schema_Object* fields_object = Schema_GetComponentUpdateFields(component_update.schema_type);
+    Schema_AddInt32(fields_object, 1, 1234);
+
+    Cast<USpatialNetDriver>(World->GetNetDriver())->Connection->SendComponentUpdate(target_entity_id, component_update);
+}
+```
+
+You could serialize and send a command request in user code in the following way:
+
+```
+Worker_RequestId SendSomeCommandRequest(Worker_EntityId target_entity_id, Worker_ComponentId component_id, uint32_t command_id, char* player_name) {
+    Worker_CommandRequest command_request;
+    command_request.component_id = component_id;
+    command_request.schema_type = Schema_CreateCommandRequest(component_id, command_id);
+    Schema_Object* request_object = Schema_GetCommandRequestObject(command_request.schema_type);
+
+    // Command specific serialization
+    Schema_AddBytes(request_object, 1, (const uint8_t*)player_name, sizeof(char) * strlen(player_name));
+
+    Worker_RequestId SendCommandRequest(target_entity_id, command_request, command_id);
+}
+```
+
+#### Receiving data
+
+You could receive and deserialize a component update and command request in user code in the following way:
+
+```
+UCLASS()
+class SPATIALGDK_API SessionComponentCallbacks : UOpCallbackTemplate
+{
+    GENERATED_BODY()
+
+public:
+    Worker_ComponentId GetComponentId() {
+        return 1337;
+    }
+
+    void OnComponentUpdate(const Worker_ComponentUpdateOp& op) override {
+        Schema_Object* ComponentObject = Schema_GetComponentUpdateFields(op.schema_type);
+
+        uint32_t player_count;
+        if (Schema_GetObjectCount(ComponentObject, component_id) > 0)
+        {
+            if (Schema_GetInt32Count(ComponentObject, 1)) {
+                player_count = Schema_GetInt32(ComponentObject, 1);
+            }
+        }
+
+        // do something with player_count
```

Comment:
As per our conversation for how to use the UWorld*. Maybe put small example here that you can spawn an actor and deserialize stuff.